tests/cmdline/commands/test_code.py

```diff
@@ -276,3 +231,54 @@ def setUp(self):
     def test_code_list_no_codes_error_message(self):
         result = self.cli_runner.invoke(code_list)
         self.assertEqual(1, result.output.count('# No codes found matching the specified criteria.'))
+
+
+@pytest.mark.parametrize('non_interactive_editor', ('sleep 1; vim -cwq',), indirect=True)
+def test_interactive_remote(clear_database_before_test, aiida_localhost, non_interactive_editor):
+    """Test interactive remote code setup."""
+    label = 'interactive_remote'
+    user_input = '\n'.join([label, 'description', 'arithmetic.add', 'yes', aiida_localhost.name, '/remote/abs/path'])
+    result = CliRunner().invoke(setup_code, input=user_input)
+    assert result.exception is None
+    assert isinstance(orm.Code.get_from_string('{}@{}'.format(label, aiida_localhost.name)), orm.Code)
+
+
+@pytest.mark.parametrize('non_interactive_editor', ('sleep 1; vim -cwq',), indirect=True)
+def test_interactive_upload(clear_database_before_test, aiida_localhost, non_interactive_editor):
+    """Test interactive code setup."""
+    label = 'interactive_upload'
+    dirname = os.path.dirname(__file__)
+    basename = os.path.basename(__file__)
+    user_input = '\n'.join([label, 'description', 'arithmetic.add', 'no', dirname, basename])
+    result = CliRunner().invoke(setup_code, input=user_input)
+    assert result.exception is None
+    assert isinstance(orm.Code.get_from_string('{}'.format(label)), orm.Code)
+
+
+@pytest.mark.parametrize('non_interactive_editor', ('sleep 1; vim -cwq',), indirect=True)
+def test_mixed(clear_database_before_test, aiida_localhost, non_interactive_editor):
+    """Test mixed (interactive/from config) code setup."""
+    from aiida.orm import Code
+    label = 'mixed_remote'
+    options = ['--description=description', '--on-computer', '--remote-abs-path=/remote/abs/path']
+    user_input = '\n'.join([label, 'arithmetic.add', aiida_localhost.name])
+    result = CliRunner().invoke(setup_code, options, input=user_input)
+    assert result.exception is None
+    assert isinstance(Code.get_from_string('{}@{}'.format(label, aiida_localhost.name)), Code)
+
+
+@pytest.mark.skip('This test is hanging for some reason.')
```

Comment:
I tried fixing it but I had even trouble understanding what the problem was. The problem was that the test hangs and you get no output. By changing the `click` source and setting default of `resilient_parsing=False` for the `Context` at least it printed a warning. Apparently the `CodeBuilder` was flagging an issue saying that properties were specified for a "remote" code even though it thinks that it concerns a local one. This is weird, because in the test what we are duplicating is a "remote" code. Anyway, I find the code of the `CodeBuilder` a bit hard to follow and I could not figure out why now all of a sudden it was complaining when I am just switching to a different manner of running the test and the test itself remaining essentially unchanged. Would be great if you may have some pointers, since I think you are the original author of the `CodeBuilder`.

Reply:
> since I think you are the original author of the CodeBuilder

I think this honor belongs to Rico, in 6873af9 .
However perhaps the issue is indeed in the code I added for code duplication?
E.g. the `Code.get_code_spec` function?
https://github.com/aiidateam/aiida-core/blob/0644e7d9ff492fdb2c94d608c575c697daee6d32/aiida/orm/utils/builders/code.py#L88-L115

Anyhow, since Codes are now Nodes, perhaps this function is basically obsolete and one could simply clone the Code?=

Reply:
Replacing this with a `clone` is probably still a valid point, perhaps open an issue for this if you agree?

Reply:
I was going to, but it does quite a bit more then just providing the functionality that `clone` does. It also provides all the validation and a way for the contextual defaults to be set in the CLI, so I am not sure that we can actually get rid of it.